fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java

```diff
@@ -0,0 +1,359 @@
+                }
+            } else { // active
+                // Worker threads are active, so sleep for a bit before we check again
+                safeSleep(5000);
+            }
+        }
+    }
+
+    /**
+     * Call the FHIR server $retrieve-index operation.
+     * @return true if the call was successful (200 OK) and index IDs were found, otherwise false
+     */
+    private boolean callRetrieveIndex() {
+        boolean result = false;
+
+        Builder builder = Parameters.builder();
+        builder.parameter(Parameter.builder().name(str(COUNT_PARAM)).value(intValue(MAX_RETRIEVE_COUNT)).build());
+        builder.parameter(Parameter.builder().name(str(NOT_MODIFIED_AFTER_PARAM)).value(str(reindexTimestamp)).build());
+        if (lastIndexId != null) {
+            builder.parameter(Parameter.builder().name(str(AFTER_INDEX_ID_PARAM)).value(str(lastIndexId)).build());
+        }
+        Parameters parameters = builder.build();
+        String requestBody = FHIRBucketClientUtil.resourceToString(parameters);
+
+        // Get index IDs of resources available to be reindexed
+        long start = System.nanoTime();
+        FhirServerResponse response = fhirClient.post(RETRIEVE_INDEX_URL, requestBody);
+        long end = System.nanoTime();
+
+        double elapsed = (end - start) / 1e9;
+        logger.info(String.format("called $retrieve-index: %d %s [took %5.3f s]", response.getStatusCode(), response.getStatusMessage(), elapsed));
+
+        if (response.getStatusCode() == HttpStatus.SC_OK) {
+            Resource resource = response.getResource();
+            if (resource != null) {
+                if (resource.is(Parameters.class)) {
+                    // Check the result to see if we should keep running
+                    result = extractIndexIds((Parameters) resource);
+                    if (!result) {
+                        logger.info("No more index IDs to retrieve");
+                    }
+                } else {
+                    logger.severe("FHIR Server retrieve-index response is not an Parameters: " + response.getStatusCode() + " " + response.getStatusMessage());
+                    logger.severe("Actual response: " + FHIRBucketClientUtil.resourceToString(resource));
+                }
+            } else {
+                // This would be a bit weird
+                logger.severe("FHIR Server retrieve-index operation returned no Parameters: " + response.getStatusCode() + " " + response.getStatusMessage());
+            }
+        } else {
+            // Stop as soon as we hit an error
+            logger.severe("FHIR Server retrieve-index operation returned an error: " + response.getStatusCode() + " " + response.getStatusMessage());
+        }
+
+        return result;
+    }
+
+    /**
+     * Extract the index IDs from the retrieve-index operation output.
+     * @param output the retrieve-index operation output
+     * @return true if index IDs were found, otherwise false
```

Comment:
Not sure this return comment lines up with the logic in the method. Not sure I understand how queued and active are working together here. I'd like to walk through this. I'd like to understand what happens when ids are rejected by the blockingQueue.

Reply:
If the IDs are rejected by the blockingQueue after waiting up to 10 seconds, then "queued = false", so `lastIndexId` is not updated.  So the next time retrieveIndex is called it calls right where it last left off.  I can change OFFER_TIMEOUT_IN_SEC to a bigger number (i.e. 30 seconds), I just didn't want it too high so if signalStop() is called while the queue is full, it doesn't take so long to shut everything it down.

Reply:
The return true, is just for if "active" is false (i.e. the worker threads are not active and consuming work), we don't want to just keep trying to enqueue more work, but rather return back to the monitorLoop() to start up the worker threads (or shutdown if running is false).